pkg/policy/l4.go

```diff
@@ -649,18 +649,19 @@ func (l4 *L4Filter) cacheFQDNSelector(sel api.FQDNSelector, selectorCache *Selec
 
 // add L7 rules for all endpoints in the L7DataMap
 func (l7 L7DataMap) addPolicyForSelector(rules *api.L7Rules, terminatingTLS, originatingTLS *TLSContext, auth *api.Auth, deny bool, sni []string, forceRedirect bool) {
-	l7policy := &PerSelectorPolicy{
-		TerminatingTLS: terminatingTLS,
-		OriginatingTLS: originatingTLS,
-		Auth:           auth,
-		IsDeny:         deny,
-		ServerNames:    NewStringSet(sni),
-		isRedirect:     !deny && (forceRedirect || terminatingTLS != nil || originatingTLS != nil || len(sni) > 0 || !rules.IsEmpty()),
-	}
-	if rules != nil {
-		l7policy.L7Rules = *rules
-	}
+	isRedirect := !deny && (forceRedirect || terminatingTLS != nil || originatingTLS != nil || len(sni) > 0 || !rules.IsEmpty())
 	for epsel := range l7 {
+		l7policy := &PerSelectorPolicy{
+			TerminatingTLS: terminatingTLS,
+			OriginatingTLS: originatingTLS,
+			Auth:           auth,
```

Comment:
What do we think about creating deep copies of `terminatingTLS`, `originatingTLS`, and `auth` here? It could help to prevent similar unexpected coupling due to mutation of these elsewhere.

Reply:
Rule merging logic does not need to modify any of the deeper structures, so this is not necessary. For these fields this is pretty fundamental as there is no way of merging different TLS contexts or auth types, so merging is limited to cases where only one of the filters has some of these, or they are already the same (DeepEqual).

Another approach we may want to look at would be to keep shared structures here, and create copies in the merging code where the modifications are needed so that the code for creating a (deep) copy and the modification of them would be in the same place. Decided to create individual copies here instead to make backporting easier.

So, while deep copying would be safer against future regressions, IMO we do not need to pay the cost of that now, and instead should change to take copies in the merging logic whenever a change is needed. But IMO this is for another PR for the main branch only.